lib/archive/unpack_unix.go

```diff
@@ -202,6 +202,56 @@ func streamCopy(op trace.Operation, stdin io.WriteCloser, tarStream io.Reader) e
 	}
 }
 
+func DockerUnpack(op trace.Operation, root string, tarStream io.Reader) (int64, error) {
+	fi, err := os.Stat(root)
+	if err != nil {
+		// the target unpack path does not exist. We should not get here.
+		op.Errorf("tar unpack target does not exist: %s", root)
+		return 0, err
+	}
+
+	if !fi.IsDir() {
+		err := fmt.Errorf("unpack root target is not a directory: %s", root)
+		op.Error(err)
+		return 0, err
+	}
+
+	// #nosec: 193 applianceBinaryPath is a constant, not a variable
+	cmd := exec.Command(string(applianceBinaryPath), op.ID(), root)
+
+	stdin, err := cmd.StdinPipe()
+
+	if err != nil {
+		op.Error(err)
+		return 0, err
+	}
+
+	if stdin == nil {
+		err = errors.New("stdin was nil")
+		op.Error(err)
+		return 0, err
+	}
+
+	var n int64
+	go func(n int64) {
+
+		defer stdin.Close()
+		if n, err = io.Copy(stdin, tarStream); err != nil {
+			op.Errorf("Error copying tarStream: %s", err.Error())
+		}
+
+	}(n)
+
+	if err = cmd.Start(); err != nil {
+		return 0, err
+	}
+
+	if err = cmd.Wait(); err != nil {
+		return 0, err
+	}
+	return n, nil
```

Comment:
you are also potentially racing here for than n value when cmd.Wait completes if cmd.Wait internals are not synchronized.

Reply:
yeah, i'll sync w/ a channel